util/run.bzl

```diff
@@ -0,0 +1,199 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Rules to run a command inside a container, and either commit the result
+to new container image, or extract specified targets to a directory on
+the host machine.
+
+Both return the image id of the committed container.
+"""
+
+load(
+    "@io_bazel_rules_docker//container:bundle.bzl",
+    "container_bundle",
+)
+
+def _extract_impl(ctx):
+    # Since we're always bundling/renaming the image in the macro, this is valid.
+    load_statement = 'docker load -i %s' % ctx.file.image_tar.short_path
+
+    # Generate a shell script to execute the run statement
+    ctx.actions.expand_template(
+        template=ctx.file._extract_tpl,
+        output=ctx.outputs.executable,
+        substitutions={
+          "%{load_statement}": load_statement,
+          "%{flags}": " ".join(ctx.attr.flags),
+          "%{image}": ctx.attr.image_name,
+          "%{command}": _process_command(ctx.attr.command),
+          "%{extract_file}": ctx.attr.extract_file,
+          "%{target}": ctx.attr.target,
+        },
+        is_executable=True,
+    )
+
+    return struct(runfiles=ctx.runfiles(files = [
+            ctx.executable.image_tar,
+            ctx.file.image_tar] + 
+            ctx.attr.image_tar.files.to_list() + 
+            ctx.attr.image_tar.data_runfiles.files.to_list()
+        ),
+    )
+
+def _commit_impl(ctx):
+    # Since we're always bundling/renaming the image in the macro, this is valid.
+    load_statement = 'docker load -i %s' % ctx.file.image_tar.short_path
+
+    # Generate a shell script to execute the run statement
+    ctx.actions.expand_template(
+        template=ctx.file._run_tpl,
+        output=ctx.outputs.executable,
+        substitutions={
+          "%{load_statement}": load_statement,
+          "%{flags}": " ".join(ctx.attr.flags),
+          "%{image}": ctx.attr.image_name,
+          "%{original_image}": ctx.attr.original_image,
+          "%{command}": _process_command(ctx.attr.command),
+        },
+        is_executable=True,
+    )
+
+    return struct(runfiles=ctx.runfiles(files = [
+            ctx.executable.image_tar,
+            ctx.file.image_tar] + 
+            ctx.attr.image_tar.files.to_list() + 
+            ctx.attr.image_tar.data_runfiles.files.to_list()
+        ),
+    )
+
+_run_and_commit = rule(
+    attrs = {
+        "flags": attr.string_list(
+            doc = "list of flags to pass to run command",
+            default = [],
+        ),
+        "image_tar": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            single_file = True,
+            cfg = "target",
+        ),
+        "original_image": attr.string(
```

Comment:
This can  be determined.
https://github.com/GoogleCloudPlatform/base-images-docker/blob/master/package_managers/download_pkgs.bzl#L22

Reply:
The input doesn't always have to be from a `docker_build` rule here though. For example, in my test I pass in a `container_pull`ed image referenced as `@python_base//image`. I think we need to pass this in as a param still.